java/src/jmri/jmrit/logixng/actions/swing/ActionSignalHeadSwing.java

```diff
@@ -39,80 +39,90 @@
     private JTextField _signalHeadOperationFormulaTextField;
-    
+
     private JTabbedPane _tabbedPaneAppearanceType;
     private JPanel _panelAppearanceTypeDirect;
     private JPanel _panelAppearanceTypeReference;
     private JPanel _panelAppearanceTypeLocalVariable;
     private JPanel _panelAppearanceTypeFormula;
-    
+
     private JComboBox<SignalHeadAppearance> _signalHeadAppearanceComboBox;
     private JTextField _signalHeadAppearanceReferenceTextField;
     private JTextField _signalHeadAppearanceLocalVariableTextField;
     private JTextField _signalHeadAppearanceFormulaTextField;
-    
+
     private BeanSelectPanel<SignalHead> _exampleSignalHeadBeanPanel;
-    
-    
+
+
     @Override
     protected void createPanel(@CheckForNull Base object, @Nonnull JPanel buttonPanel) {
         ActionSignalHead action = (ActionSignalHead)object;
-        
+
         panel = new JPanel();
         panel.setLayout(new BoxLayout(panel, BoxLayout.Y_AXIS));
-        
+
         JPanel examplePanel = new JPanel();
         JPanel innerExamplePanel = new JPanel();
         innerExamplePanel.setBorder(BorderFactory.createLineBorder(Color.black));
         _exampleSignalHeadBeanPanel = new BeanSelectPanel<>(InstanceManager.getDefault(SignalHeadManager.class), null);
         innerExamplePanel.add(_exampleSignalHeadBeanPanel);
-        
+
+        _exampleSignalHeadBeanPanel.getBeanCombo().addActionListener((java.awt.event.ActionEvent e) -> {
+            setAppearanceComboBox(null);
+        });
+
+
         JPanel actionPanel = new JPanel();
-        
-        
+
+
         // Set up tabbed pane for selecting the signal head
         _tabbedPaneSignalHead = new JTabbedPane();
         _panelSignalHeadDirect = new javax.swing.JPanel();
         _panelSignalHeadReference = new javax.swing.JPanel();
         _panelSignalHeadLocalVariable = new javax.swing.JPanel();
         _panelSignalHeadFormula = new javax.swing.JPanel();
-        
+
         _tabbedPaneSignalHead.addTab(NamedBeanAddressing.Direct.toString(), _panelSignalHeadDirect);
         _tabbedPaneSignalHead.addTab(NamedBeanAddressing.Reference.toString(), _panelSignalHeadReference);
         _tabbedPaneSignalHead.addTab(NamedBeanAddressing.LocalVariable.toString(), _panelSignalHeadLocalVariable);
         _tabbedPaneSignalHead.addTab(NamedBeanAddressing.Formula.toString(), _panelSignalHeadFormula);
-        
+
         _tabbedPaneSignalHead.addChangeListener((ChangeEvent e) -> {
-            enableDisableExampleSignalHeadBeanPanel();
+            setGuiEnabledStates();
         });
-        
+
         _signalHeadBeanPanel = new BeanSelectPanel<>(InstanceManager.getDefault(SignalHeadManager.class), null);
         _panelSignalHeadDirect.add(_signalHeadBeanPanel);
-        
+
+        _signalHeadBeanPanel.getBeanCombo().addActionListener((java.awt.event.ActionEvent e) -> {
+            setAppearanceComboBox(null);
+        });
```

Comment:
> > Great! Thanks!
> > When I test this PR and edit a SignalHead action, the appearance is wrong when the swing dialog opens. I can select a signal head and change the appearance and that works fine. But when I edit the SignalHead again, it shows the first appearance, which in my case is Dark. I have the same error with the SignalMast action, and with the SignalHead and SignalMast expressions.
> 
> I cannot re-create your scenario.

It seems to be a timing issue. If I comment out line 109, the problem disappears.

The problem seems to go away if you change line 271 to:
`jmri.util.ThreadingUtil.runOnGUIEventually(() -> { setAppearanceComboBox(action); });`